lib/ecto/repo/preloader.ex

```diff
@@ -316,16 +316,75 @@ defmodule Ecto.Repo.Preloader do
     """
 
   defp preload_order(assoc, query, related_field) do
-    custom_order_by = Enum.map(assoc.preload_order, fn
-      {direction, field} ->
-        {direction, related_key_to_field(query, {0, field})}
-      field ->
-        {:asc, related_key_to_field(query, {0, field})}
-    end)
+    preload_order = resolve_preload_order(assoc.preload_order)
+
+    custom_order_by =
+      Enum.map(preload_order, fn
+        {direction, %Ecto.Query.DynamicExpr{} = dynamic} ->
+          unless Ecto.Query.Builder.OrderBy.valid_direction?(direction) do
+            raise ArgumentError,
+                  "`:preload_order` must specify valid directions, " <>
+                    "got: `#{inspect(preload_order)}`, `#{inspect(direction)}` is not a valid direction"
+          end
+
+          {direction, dynamic_to_field(query, dynamic)}
+
+        {direction, field} ->
+          {direction, related_key_to_field(query, {0, field})}
+
+        %Ecto.Query.DynamicExpr{} = dynamic ->
+          {:asc, dynamic_to_field(query, dynamic)}
+
+        field when is_atom(field) ->
+          {:asc, related_key_to_field(query, {0, field})}
+
+        other ->
+          raise ArgumentError,
+                "`:preload_order` must resolve to a keyword list or a list of atoms/fields, " <>
+                  "got: `#{inspect(preload_order)}`, `#{inspect(other)}` is not valid"
+      end)
 
     [{:asc, related_field} | custom_order_by]
   end
 
+  defp resolve_preload_order(order) when is_list(order), do: order
+
+  defp resolve_preload_order({m, f, a}) do
+    case apply(m, f, a) do
+      order when is_list(order) ->
+        order
+
+      other ->
+        raise ArgumentError,
+              "`:preload_order` must resolve to a keyword list or a list of atoms/fields, " <>
+                "got: `#{inspect(other)}`"
+    end
+  end
+
+  defp dynamic_to_field(query, dynamic) do
+    unless length(dynamic.binding) == Ecto.Query.Builder.count_binds(query) do
+      raise ArgumentError,
+            """
+            dynamic expressions inside `:preload_order` must have the correct number of bindings,
+            got: `#{inspect(dynamic)}` for query:
+
+            #{Inspect.Ecto.Query.to_string(query)}
+            """
+    end
+
+    {expr, _, _, _, _, _} = Ecto.Query.Builder.Dynamic.fully_expand(%Ecto.Query{}, dynamic)
+
+    case expr do
+      {{:., _, [{:&, _, [_ix]}, _field]}, [], []} = field_ast ->
+        field_ast
```

Comment:
Do we need to do this validation? Could we simply return the `dynamic` field instead?

Reply:
I think we have to but this was one of the more difficult parts for me. I could let you know my rationale:

The existing `preload_order` behaviour is to [insert the order before any of the `order_by`s that are already in the preload query](https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/repo/preloader.ex#L264):

```elixir
  update_in query.order_bys, fn order_bys ->
    [%Ecto.Query.QueryExpr{expr: preload_order(assoc, query, field), params: [],
                           file: __ENV__.file, line: __ENV__.line}|order_bys]
```

So because of that it basically creates the AST that would normally be created by the builder and injects it. If don't expand the dynamic then it stays as `%DynamicExpr{}`.

Reply:
Something similar happens for the keyword list version of this as well only it's a lot simpler to achieve it https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/repo/preloader.ex#L329

```elixir
 defp preload_order(assoc, query, related_field) do
  custom_order_by = Enum.map(assoc.preload_order, fn
    {direction, field} ->
      {direction, related_key_to_field(query, {0, field})}
    field ->
      {:asc, related_key_to_field(query, {0, field})}
  end)

  [{:asc, related_field} | custom_order_by]
end

defp related_key_to_field(query, {pos, key}) do
  {{:., [], [{:&, [], [related_key_pos(query, pos)]}, key]}, [], []}
end
```

Reply:
I am just worried this will fail in unpredictable ways. Maybe we can add a fourth argument  to `order_by` which would allow us to prepend? We can also use reverse_order + append + reverse_order. But probably too much work for not much. Thoughts?

Reply:
Oh I actually like both those ideas. The 4th argument is a bit better IMO because it gives us and other users extra functionality  and is a bit clearer when reading the code. I could implement that if you agree?

Reply:
Please go ahead as you prefer!